app/src/main/res/layout/fragment_device_list.xml

```diff
@@ -0,0 +1,119 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="15dp"
+        android:layout_marginStart="15dp"
+        android:paddingLeft="5dp"
+        android:paddingTop="5dp"
+        android:paddingRight="5dp"
+        android:paddingBottom="1dp"
+        android:text="@string/your_device"
+        android:textSize="13sp"
+        android:textStyle="italic"
+        app:layout_constraintHorizontal_bias="0.0"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <TextView
+        android:id="@+id/text_view_device_name"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:layout_constraintHorizontal_bias="0.0"
+        app:layout_constraintTop_toBottomOf="@+id/text_view_your_device"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        android:hint="@string/device_name"
+        android:textStyle="bold"
+        android:textSize="17sp"
+        android:layout_marginLeft="15dp"
+        android:layout_marginStart="15dp"
+        android:paddingLeft="5dp"
+        android:paddingRight="5dp"/>
+
+    <TextView
+        android:id="@+id/text_view_device_status"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:layout_constraintHorizontal_bias="0.0"
+        app:layout_constraintTop_toBottomOf="@+id/text_view_device_name"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        android:hint="@string/device_status"
+        android:textSize="15sp"
+        android:layout_marginLeft="15dp"
+        android:layout_marginStart="15dp"
+        android:paddingLeft="5dp"
+        android:paddingRight="5dp"
+        android:paddingBottom="5dp"/>
+
+    <View
+        android:id="@+id/view_device_list_boundary"
+        android:layout_width="match_parent"
+        android:layout_height="2dp"
+        android:background="#0000f0"
+        app:layout_constraintTop_toBottomOf="@+id/text_view_device_status"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        android:layout_marginLeft="5dp"
+        android:layout_marginRight="5dp"/>
+
+    <TextView
+        android:id="@+id/text_view_available_device"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/nearby_devices"
+        android:gravity="center"
+        android:textSize="16sp"
+        android:fontFamily="monospace"
+        app:layout_constraintTop_toBottomOf="@+id/view_device_list_boundary"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        android:paddingTop="5dp"/>
+
+    <FrameLayout
```

Comment:
I would say this layout is not needed, the point of constraintlayout is for a completely flat hierarchy, if you cannot achieve that then something has gone wrong.
You can use a [group](https://developer.android.com/reference/android/support/constraint/Group) to control the visibility of many views at once

Reply:
Based on what I had read, ListFragment requires both a list view and an optional text view (to be displayed in case the list is empty) to be kept in it's FrameLayout, which is why the nesting

Reply:
```
Optionally, your view hierarchy can contain another view object of any type to display when the list view is empty. This "empty list" notifier must have an id "android:empty".
```
Sounds like the view can be anywhere in the hierarchy. https://developer.android.com/reference/android/app/ListFragment

Reply:
I'll try keeping both outside then

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/e18dc725b8f7758b6243d8ae5b58bcdf73d74a15)
I tried using a `group` but it was accepting only one of the _list_ and _empty_ ids, probably because they are system defined. (Accepted multiple user defined ids just fine).
So I'm changing their visibility individually now.